posts/inside-rust/2022-11-02-async-fn-in-trait-nightly.md

```diff
@@ -0,0 +1,273 @@
+---
+layout: post
+title: "Async fn in trait MVP comes to nightly"
+author: Tyler Mandry
+team: The Rust Async Working Group <https://www.rust-lang.org/governance/wgs/wg-async>
+---
+
+`async` and `.await` were a major improvement in the ergonomics of writing async code in Rust. If you use async Rust for long, however, you'll surely discover a big limitation: `async fn` doesn't work in traits.
```

Comment:
I find the structure of this blog post really buries the lede. As a user, I want to know what is working, what it does, and how to use it, and I think that should come first in the blog post and summarised in the first sentence. Instead the post starts on a tangent and I have to wade through a bunch of compiler internals discussion before getting to the interesting bit. I think the discussion is interesting and clearly explained, but I think it should come after introducing the feature and the limitations discussion.

The tone could also be a bit more celebratory - at the moment phrasing like "you'll surely discover a big limitation" and "How, you might reasonably ask, could we ship a feature that doesn't work in *traits*?" sound pretty negative, when this blog post is announcing something really cool which is the culmination of a load of work.

Reply:
Hmm, this is a good point, I tend to agree about burying the lede -- though I do think it's important to temper expectations. We are not announcing that async fn in traits works. We are saying that an important subset of the feature is available on nightly, and talking about the next steps that come after that.

Reply:
This suggestion doesn't quite flow into the next paragraph, but maybe something like:

```suggestion
`async` and `.await` were a major improvement in the ergonomics of writing async code in Rust. If you use async Rust for long, however, you'll surely discover a big limitation: `async fn` doesn't work in traits.  We've just released an MVP that starts to close this gap and is ready for wider testing.  Read on to see what works today and what limitations still exist.
```

Reply:
Thanks for this feedback. I've revised the post to put the initial "good news" front and center, but kept the background information first after the introduction.